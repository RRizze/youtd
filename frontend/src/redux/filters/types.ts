import { ActionCreator } from "redux";

export interface SelectFilterState {
  select: boolean;
  not: boolean;
}

export interface RangeFilterState {
  min: number;
  max: number;
}

export interface Rarity {
  all: boolean;
  common: SelectFilterState;
  uncommon: SelectFilterState;
  rare: SelectFilterState;
  unique: SelectFilterState;
}

export interface Elements {
  all: boolean;
  nature: SelectFilterState;
  fire: SelectFilterState;
  ice: SelectFilterState;
  astral: SelectFilterState;
  darkness: SelectFilterState;
  storm: SelectFilterState;
  iron: SelectFilterState;
}

export interface AttackTypes {
  all: boolean;
  elemental: SelectFilterState;
  physical: SelectFilterState;
  energy: SelectFilterState;
  decay: SelectFilterState;
  magic: SelectFilterState;
  essence: SelectFilterState;
}

export interface SelectFilters {
  rarity: Rarity;
  elements: Elements;
  attack_types: AttackTypes;
}

export type SelectFiltersUnion = Rarity | Elements | AttackTypes;

export interface RangeFilters {
  damage: RangeFilterState;
  range: RangeFilterState;
  attack_cd: RangeFilterState;
  element_lvl_up: RangeFilterState;
  wave_lvl_up: RangeFilterState;
  goldcost: RangeFilterState;
}

export interface FiltersState {
  count_rows: number;
  offset_rows: number;
  select: SelectFilters;
  ranges: RangeFilters;
};
