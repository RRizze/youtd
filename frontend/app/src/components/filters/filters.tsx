import React from 'react';
import { FlexBox } from '@ui/flex-box';
import { IProps } from './filters.types';

const Filters = ({ children }: IProps) => (
  <FlexBox
    justifyContent='flex-start'
    flexDirection='column'
    alignItems='flex-start'
    pb='20px'
    pt='20px'
  >
    {children}
  </FlexBox>
);

export default Filters;
