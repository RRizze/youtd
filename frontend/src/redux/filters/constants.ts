export const SET_SELECT_FILTER = "SET_SELECT_FILTER";
export const RESET_SELECT_FILTER = "RESET_SELECT_FILTER";
export const SET_RANGE_VALUE = "SET_RANGE_VALUE";
export const RESET_COUNT_ROWS = "RESET_COUNT_ROWS";
export const SET_COUNT_ROWS = "SET_COUNT_ROWS";
export const SET_NOT_FILTER_SELECT = "SET_NOT_FILTER_SELECT";
