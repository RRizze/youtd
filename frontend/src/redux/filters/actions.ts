import * as constants from "./constants";
import { SelectFiltersUnion } from "./types";
import { createAction, ActionsUnion } from "../types";

export const setFilterAll = (
  filter: string,
  filters: SelectFiltersUnion,
  not?: string
) => {
  let newFilters;
  if (not !== "not") {
    newFilters = {
      ...filters,
      [filter]: {
        select: !filters[filter].select,
        not: false,
      },
    };
  } else {
    newFilters = {
      ...filters,
      [filter]: {
        select: false,
        not: !filters[filter].not,
      },
    };
  }

  for (const filter in newFilters) {
    if (
      (newFilters[filter]["select"] || newFilters[filter]["not"])
      && newFilters[filter] !== "all"
    ) {
      return false;
    }
  }
  return true;
};

export const setNotSelectFilter = (
  filter: string,
  filterType: string,
  filters: SelectFiltersUnion
) => createAction(constants.SET_NOT_FILTER_SELECT, {
  filter,
  filterType,
  all: setFilterAll(filter, filters, "not"),
});

export const setSelectFilter = (
  filter: string,
  filterType: string,
  filters: SelectFiltersUnion
) => createAction(constants.SET_SELECT_FILTER, {
  filter,
  filterType,
  all: setFilterAll(filter, filters),
});

export const resetSelectFilter = (filterType: string) => createAction(constants.RESET_SELECT_FILTER, {
  filterType,
});

export const setCountRows = () => createAction(constants.SET_COUNT_ROWS);
export const resetCountRows = () => createAction(constants.RESET_COUNT_ROWS);

export const setRangeFilter = (
  min: number,
  max: number,
  filterType: string
) => createAction(constants.SET_RANGE_VALUE, {
  min,
  max,
  filterType,
});

const Actions = {
  setSelectFilter,
  resetSelectFilter,
  setCountRows,
  resetCountRows,
  setNotSelectFilter,
  setRangeFilter,
};

export type Actions = ActionsUnion<typeof Actions>;
