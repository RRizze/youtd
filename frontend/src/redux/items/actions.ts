import * as constants from "./constants";
import { normalize, createUrl } from "../../utils/helpers";
import { IItemNORM } from "../../models/item";
import { createAction, ActionsUnion } from "../types";
import { FiltersState } from "../filters/types";

export const loadItemsRequest = () => createAction(constants.LOAD_ITEMS_REQUEST);

export const loadItemsSuccess = (items: IItemNORM[], count: number) => createAction(constants.LOAD_ITEMS_SUCCESS, {
  items,
  count,
});

export const clearItemsList = () => createAction(constants.CLEAR_ITEMS_LIST);

export const loadItemsFailure = () => createAction(constants.LOAD_ITEMS_FAILURE, new Error("error"), true);

export const fetchItems = (filters: FiltersState, clear: boolean) => async (dispatch) => {
  if (clear) {
    await dispatch(clearItemsList());
  }
  await dispatch(loadItemsRequest());
  try {
    const url = await createUrl(filters, "items");
    const response = await fetch(url);
    const json = await response.json();
    const norm = await normalize(json);
    dispatch(loadItemsSuccess(norm[0], norm[1][0].count));
  } catch (error) {
    dispatch(loadItemsFailure);
  }
};

const Actions = {
  loadItemsRequest,
  loadItemsSuccess,
  loadItemsFailure,
  clearItemsList,
};

export type Actions = ActionsUnion<typeof Actions>;
