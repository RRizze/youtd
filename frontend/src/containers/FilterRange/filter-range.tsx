import React, { createRef } from "react";
import { connect } from "react-redux";
import "./filter-range-style.scss";
import { map } from "../../utils/helpers";
import { setRangeFilter } from "../../redux/filters/actions";

export interface Props {
  title: string;
  filterType: string;
  min: number;
  max: number;
  setRangeFilter: typeof setRangeFilter;
}

export interface State {
  rangeMin: number;
  rangeMax: number;
  sliderWidth: number;
  shiftXLower: number;
  shiftXUpper: number;
  shiftXLowerStart: number;
  shiftXUpperStart: number;
  translateX: number;
  scaleX: number;
}

const initialState = {
  rangeMin: 0,
  rangeMax: 1,
  sliderWidth: 220,
  shiftXLower: 0,
  shiftXUpper: 184,
  shiftXLowerStart: 0,
  shiftXUpperStart: 0,
  translateX: 0,
  scaleX: 1,
};

export class FilterRange extends React.Component<Props, State> {
  state = initialState;

  lowerRef = createRef<HTMLDivElement>();
  upperRef = createRef<HTMLDivElement>();
  sliderRef = createRef<HTMLDivElement>();

  // Max
  handleMouseMoveMax = (e) => {
    e.preventDefault();

    const rightEdge = this.sliderRef.current.offsetWidth - this.upperRef.current.offsetWidth * 2;
    let newLeftMax = e.pageX -
      this.state.shiftXUpperStart -
      this.sliderRef.current.getBoundingClientRect().left -
      this.upperRef.current.offsetWidth;

    if (newLeftMax < 0) {
      newLeftMax = 0;
    }

    if (newLeftMax > rightEdge) {
      newLeftMax = rightEdge;
    }

    if(newLeftMax < this.state.shiftXLower) {
      newLeftMax = this.state.shiftXLower;
    }

    this.setState((prevState) => ({
      shiftXUpper: newLeftMax,
      translateX: (prevState.shiftXLower - (184 - newLeftMax)) / 2,
      scaleX: 1 - 0.005 * (184 - newLeftMax) - (0.005 * prevState.shiftXLower),
    }));
  }

  handleMouseDownMax = (e) => {
    e.preventDefault();

    this.setState({
      shiftXUpperStart: e.pageX - this.upperRef.current.getBoundingClientRect().left,
    });

    document.addEventListener("mousemove", this.handleMouseMoveMax);
    document.addEventListener("mouseup", this.handleMouseUpMax);
  }

  handleMouseUpMax = (e) => {
    e.preventDefault();

    const { shiftXUpper, rangeMax, shiftXLower, rangeMin } = this.state;
    const { filterType, setRangeFilter } = this.props;

    setRangeFilter(
      map(shiftXLower, 0, 202, rangeMin, rangeMax),
      map(shiftXUpper, 0, 184, rangeMin, rangeMax),
      filterType);

    document.removeEventListener("mousemove", this.handleMouseMoveMax);
    document.removeEventListener("mouseup", this.handleMouseUpMax);
  }

  // Min
  handleMouseMove = (e) => {
    e.preventDefault();

    const rightEdge = this.sliderRef.current.offsetWidth - this.lowerRef.current.offsetWidth;
    let newLeft = e.pageX - this.state.shiftXLowerStart - this.sliderRef.current.getBoundingClientRect().left;

    if (newLeft < 0) {
      newLeft = 0;
    }

    if (newLeft > rightEdge) {
      newLeft = rightEdge;
    }

    if(newLeft > this.state.shiftXUpper) {
      newLeft = this.state.shiftXUpper;
    }

    this.setState((prevState) => ({
      shiftXLower: newLeft,
      translateX: newLeft / 2 - (184 - prevState.shiftXUpper) / 2,
      scaleX: 1 - 0.005 * newLeft - 0.005 * (184 - prevState.shiftXUpper),
    }));
  }

  handleMouseUp = (e) => { // 2 
    e.preventDefault();

    const { shiftXLower, shiftXUpper, rangeMin, rangeMax } = this.state;
    const { filterType, setRangeFilter } = this.props;

    const min = map(shiftXLower, 0, 202, rangeMin, rangeMax);
    const max = map(shiftXUpper, 0, 184, rangeMin, rangeMax);

    setRangeFilter(min, max, filterType);
    document.removeEventListener("mousemove", this.handleMouseMove);
    document.removeEventListener("mouseup", this.handleMouseUp);
  }

  handleMouseDown = (e) => { // 1
    e.preventDefault();

    this.setState({
      shiftXLowerStart: e.pageX - this.lowerRef.current.getBoundingClientRect().left,
    });

    document.addEventListener("mousemove", this.handleMouseMove);
    document.addEventListener("mouseup", this.handleMouseUp);
  }

  componentDidMount() {
    this.setState({
      rangeMin: this.props.min,
      rangeMax: this.props.max,
    });
  }

  render() {
    const min = map(this.state.shiftXLower, 0, 202, this.state.rangeMin, this.state.rangeMax);
    const max = map(this.state.shiftXUpper, 0, 184, this.state.rangeMin, this.state.rangeMax);

    return (
      <div className="filter-range">
        <div className="filter-header">
          <h2 className="filter-title">{this.props.title}</h2>
          <div className="filter-values">
            <p>{min} - {max}</p>
          </div>
        </div>
        <div
          className="filter-body"
          style={{ width: `${this.state.sliderWidth}px` }}
          ref={this.sliderRef}>
          <div className="range-line" />
          <div
            className="filter-range__line"
            style={{
              transform: `translate(${this.state.translateX}px, 0)
                scale(${this.state.scaleX}, 1)`,
            }} />
          <div
            data-value="min"
            ref={this.lowerRef}
            onMouseDown={this.handleMouseDown}
            onDragStart={() => false}
            className="filter-range__min-value"
            style={{
              transform: `translate(${this.state.shiftXLower}px, 0)`,
            }}>
          </div>
          <div
            data-value="max"
            ref={this.upperRef}
            onMouseDown={this.handleMouseDownMax}
            onDragStart={() => false}
            className="filter-range__max-value"
            style={{
              transform: `translate(${this.state.shiftXUpper}px, 0)`,
            }}>
          </div>
        </div>
      </div>
    );
  }
}

export default connect(null, { setRangeFilter })(FilterRange);
