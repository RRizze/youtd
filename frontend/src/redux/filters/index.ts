import * as constants from "./constants";
import { Actions } from "./actions";
import { FiltersState } from "./types";

export const initialState: FiltersState = {
  count_rows: 10,
  offset_rows: 0,
  select: {
    rarity: {
      all: true,
      common: {
        select: false,
        not: false,
      },
      uncommon: {
        select: false,
        not: false,
      },
      rare: {
        select: false,
        not: false,
      },
      unique: {
        select: false,
        not: false,
      },
    },
    elements: {
      all: true,
      nature: {
        select: false,
        not: false,
      },
      fire: {
        select: false,
        not: false,
      },
      ice: {
        select: false,
        not: false,
      },
      astral: {
        select: false,
        not: false,
      },
      darkness: {
        select: false,
        not: false,
      },
      storm: {
        select: false,
        not: false,
      },
      iron: {
        select: false,
        not: false,
      },
    },
    attack_types: {
      all: true,
      elemental: {
        select: false,
        not: false,
      },
      physical: {
        select: false,
        not: false,
      },
      energy: {
        select: false,
        not: false,
      },
      decay: {
        select: false,
        not: false,
      },
      magic: {
        select: false,
        not: false,
      },
      essence: {
        select: false,
        not: false,
      },
    },
  },
  ranges: {
    damage: {
      min: 1,
      max: 20000,
    },
    range: {
      min: 0,
      max: 10000,
    },
    attack_cd: {
      min: 0,
      max: 10,
    },
    element_lvl_up: {
      min: 0,
      max: 15,
    },
    wave_lvl_up: {
      min: 0,
      max: 90,
    },
    goldcost: {
      min: 0,
      max: 10000,
    },
  },
};

export default (state = initialState, action: Actions) => {
  switch (action.type) {
    case constants.RESET_COUNT_ROWS:
      return {
        ...state,
        offset_rows: 0,
      };
    case constants.SET_COUNT_ROWS:
      return {
        ...state,
        offset_rows: state.offset_rows + 10,
      };
    case constants.SET_SELECT_FILTER:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...state.select[action.payload.filterType],
            all: action.payload.all,
            [action.payload.filter]: {
              select: !state.select[action.payload.filterType][action.payload.filter].select,
              not: false,
            },
          },
        },
      };
    case constants.RESET_SELECT_FILTER:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...initialState.select[action.payload.filterType],
          },
        },
      };
    case constants.SET_RANGE_VALUE:
      return {
        ...state,
        offset_rows: 0,
        ranges: {
          ...state.ranges,
          [action.payload.filterType]: {
            min: action.payload.min,
            max: action.payload.max,
          },
        },
      };
    case constants.SET_NOT_FILTER_SELECT:
      return {
        ...state,
        offset_rows: 0,
        select: {
          ...state.select,
          [action.payload.filterType]: {
            ...state.select[action.payload.filterType],
            all: action.payload.all,
            [action.payload.filter]: {
              select: false,
              not: !state.select[action.payload.filterType][action.payload.filter].not,
            },
          },
        },
      };
    default:
      return state;
  }
};
