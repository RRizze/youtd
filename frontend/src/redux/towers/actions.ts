import {Dispatch} from "redux";
import * as constants from "./constants";
import { ITowerDTO } from "../../models/tower";
import { createAction, ActionsUnion } from "../types";
import { FiltersState } from "../filters/types";
import { createUrl } from "../../utils/helpers";

export type DispatchPromise = (dispatch: Dispatch) => Promise<Actions>;

export type FetchTowers = (filters: FiltersState, clear: boolean) => DispatchPromise;

export const loadTowersRequest = () => {
  return createAction(constants.LOAD_TOWERS_REQUEST);
};

export const clearTowersList = () => {
  return createAction(constants.CLEAR_TOWERS_LIST);
};

export const loadTowersSuccess = (towers: ITowerDTO[], count: number) => {
  return createAction(constants.LOAD_TOWERS_SUCCESS, {
    towers,
    count,
  });
};
export const loadTowersFailure = () => {
  return createAction(constants.LOAD_TOWERS_FAILURE);
};

export const fetchTowers: any = (
  filters: FiltersState,
  clear: boolean
) => async (dispatch) => {
  if (clear) {
    dispatch(clearTowersList());
  }
  dispatch(loadTowersRequest());
  try {
    const url = await createUrl(filters, "towers");
    const response = await fetch(url);
    const json = await response.json();
    return dispatch(loadTowersSuccess(json[0], json[1][0].count));
  } catch (error) {
    dispatch(loadTowersFailure());
  }
};

export const Actions = {
  loadTowersRequest,
  loadTowersSuccess,
  loadTowersFailure,
  clearTowersList,
  fetchTowers,
};

export type Actions = ActionsUnion<typeof Actions>;
