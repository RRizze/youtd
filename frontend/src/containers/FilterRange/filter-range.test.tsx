import React from "react";
import { mount } from "enzyme";
import { FilterRange } from "./filter-range";
import { setRangeFilter } from "../../redux/filters/actions";
import renderer from "react-test-renderer";

describe("Section tests", () => {
  it("render correctly", () => {
    const props = {
      title: "damage",
      filterType: "damage",
      setRangeFilter,
      min: 0,
      max: 100,
    },
      tree = renderer
        .create(<FilterRange {...props} />)
        .toJSON();
      
    expect(tree).toMatchSnapshot();
  });

  it("render with props", () => {
    const props = {
      title: "damage",
      filterType: "damage",
      setRangeFilter,
      min: 0,
      max: 100,
    },
      wrapper = mount(<FilterRange {...props} />);
    expect(wrapper.props()).toEqual(props);
  });
});
