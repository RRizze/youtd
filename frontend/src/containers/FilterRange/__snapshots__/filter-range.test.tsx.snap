// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Section tests render correctly 1`] = `
<div
  className="filter-range"
>
  <div
    className="filter-header"
  >
    <h2
      className="filter-title"
    >
      damage
    </h2>
    <div
      className="filter-values"
    >
      <p>
        0
         - 
        100
      </p>
    </div>
  </div>
  <div
    className="filter-body"
    style={
      Object {
        "width": "220px",
      }
    }
  >
    <div
      className="range-line"
    />
    <div
      className="filter-range__line"
      style={
        Object {
          "transform": "translate(0px, 0)
                scale(1, 1)",
        }
      }
    />
    <div
      className="filter-range__min-value"
      data-value="min"
      onDragStart={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "transform": "translate(0px, 0)",
        }
      }
    />
    <div
      className="filter-range__max-value"
      data-value="max"
      onDragStart={[Function]}
      onMouseDown={[Function]}
      style={
        Object {
          "transform": "translate(184px, 0)",
        }
      }
    />
  </div>
</div>
`;
